### ISSUE TITLE
Progressive enhancement

### ISSUE THREAD

**Issue by fdecampredon** [ractivejs/ractive#395]
Actually when we prerender Ractive template on server, then try to rerender them on client the dom is replaced.
Do you think it would be possible to have an options that would allow Ractive to not replace the dom but use the already built in one ?
To ensuring that the dom already present would be the same that the dom produced prehaps you could use some kind of checksum (a little alike what React guys do)

Thanks for your work anyway

**Comment by PaquitoSoft** [comment]
Just wanted to add another developer who is quite interested in this feature.

**Comment by kurdin** [comment]
Is there any progress on this?

**Comment by Rich-Harris** [comment]
@kurdin was just about to direct you here from #957 :-) No, there hasn't been any progress, but I'm anxious for us to make a start on it soon. I envision it as being the big new feature for 0.8.x.

**Comment by kurdin** [comment]
@Rich-Harris Would love to help with testing :) 

btw: I did some experiments with DOM changes after ``ractive.render()`` event and it looks like if you modify something inside ractive init element and then change the data to trigger render update, those new changes to DOM that reflects new data can be corrupted.  Is this normal? 

Here is an example: http://jsfiddle.net/msx4r0s2/1/ 
if I open dev tools (inspect element) and modify #app.h1 to remove it's content then click on radio button to trigger language change, that modified H1 will always be empty. Is it a bug or this is how ractive deal with DOM changes?

**Comment by Rich-Harris** [comment]
@kurdin watch this space :)

That behaviour you're experiencing is expected - Ractive works by storing references to DOM nodes (including text nodes within an element) and updating them when necessary. By making the assumption that the DOM *hasn't* been modified by something other than Ractive, it can make those updates in a very granular fashion that saves throwing out big chunks of perfectly good DOM.

So it's best to avoid imperative DOM manipulation (unless it's done in a way that doesn't conflict with Ractive's data binding). Though in theory, the same logic that would handle progressive enhancement could theoretically allow us to force the DOM to resync with Ractive's view of the world...

**Comment by NikhilVerma** [comment]
Sorry to bump but this is the very first thing the folks in my team asked when I proposed RactiveJs for server side rendering. Right now it's way easier for us to migrate to Ractive compared to React so we have compromised for a server render -> client render approach. But this would frankly be amazing.

**Comment by msokk** [comment]
Still planned? Currently pondering about ways using Ractive on the backend, so this capability would be much appreciated.

**Comment by hkaju** [comment]
+1 for more info on the current status of this.

**Comment by beauCielBleu** [comment]
+1 I need this.

2015-10-06 17:54 GMT+02:00 Hendrik Kaju <notifications@github.com>:

**Comment by Rich-Harris** [comment]
Me too! Top of the list as soon as 0.8 is done. It's killing me that this issue has been open for so long

**Comment by beauCielBleu** [comment]
Oh great! Thanks Rich-Harris and msokk for the good news!

**Comment by monoblaine** [comment]
I'd like to share my own solution. It's not perfect, probably far from the ideal solution (mine entirely replaces the DOM), but it is useful for me and I use it in my projects.

Basically, I decorate a container element and the input fields inside with custom `data` attributes, then the plugin dynamically creates a Ractive template, using the container element's html as a Ractive `partial`.

Here is an example:
https://jsbin.com/qucewu/edit?html,output

**Comment by Rich-Harris** [comment]
Great news! @evs-chris decided that enough was enough and went ahead and implemented progressive enhancement: http://jsfiddle.net/rich_harris/kxpdgo89/

There are a couple of pieces we still need to consider:

* Encapsulating CSS, if you're using that feature of components in a server-rendered app, is something that doesn't currently work
* Not all two-way bindings work yet – need to implement e.g. `<select>` bindings

This is really exciting stuff – it means that it's possible to send pre-rendered pages down the wire for the fastest possible perceived performance, with basically zero sacrifice in terms of interactivity or ease of development. When you get round to loading your JavaScript, Ractive will seamlessly take over. If you have a Node.js backend, this is something that will work really well with isomorphic/universal (whichever term you prefer) components, but unlike some server-side rendering setups there's no implied assumption about how the markup got to the browser – like everything in Ractive, it's designed to bend to your will.

The one simplifying assumption is this: the markup being enhanced has the same 'shape' as the template. There's no complex diffing going on – if the next element to be enhanced doesn't match the next element in the template, Ractive will just discard it and create a new one.

To use it, just add `enhance: true` to the options object when you create a new Ractive instance. You'll need to be running the edge version, which you can grab [here](http://cdn.ractivejs.org/edge/ractive.js), or install from npm with:

```bash
npm install --save git://github.com/ractivejs/ractive#build

# or, for a specific commit from https://github.com/ractivejs/ractive/commits/build:
npm install --save git://github.com/ractivejs/ractive#8779d5651aa13189d8d1d0abb001865f2c3d1a4f
```

I'm going to close this issue, but if you try this out and run into problems, please let us know!

**Comment by beauCielBleu** [comment]
Thanks a lot guys for the good work! I will test that today.

**Comment by TehShrike** [comment]
Will this be in 0.8, or later?

**Comment by Madgvox** [comment]
@TehShrike It will be in 0.8.

**Comment by TehShrike** [comment]
sweeeeet!

**Comment by fabien** [comment]
@Rich-Harris will this work with loops or partials? It would surprise me if it did ;-)

**Comment by Madgvox** [comment]
@fabien As long as it matches up to the rendered template, it [Just Works™](http://jsfiddle.net/kxpdgo89/8/).

**Comment by fabien** [comment]
@Madgvox whoah, that's amazing! Great work guys!

**Comment by kurdin** [comment]
If you look closer in chrome dev tools, you can see that on render it is touching every single element of the existed DOM tree in <main> even if content changed only in <h1> unfortunately this is not ideal solution and still using something like virtual-dom is better (in some cases). Just saying...

**Comment by Madgvox** [comment]
Ractive's progressive enhancement does not replace virtualdom. It simply binds the ractive's virtualdom to the existing DOM elements rather than creating new ones.

**Comment by martypdx** [comment]
Guessing that what @kurdin is noticing is that unlike update, render doesn't avoid touching the DOM if value is same.

Perhaps would be better anyway to compare actual DOM value than store another js value in virtual dom. This is what google is doing in new template engine (need link)

**Comment by kurdin** [comment]
@Madgvox @martypdx I am not sure how reactive changes elements but it looks like it just replace existing elements in DOM tree with newly rendered. May be I am wrong. 

This is how it should work, touching only elements with changes:
http://jsfiddle.net/d0kz1ov5/9/

**Comment by martypdx** [comment]
@kurdin Looks like it detaches then reinserts the _same_ nodes.

**Comment by evs-chris** [comment]
It shouldn't detach any nodes if the template matches. It looks like what's happening is the text content isn't quite matching (whitespace), so the text is getting updated for the `<p>` tags. An `htmlEqual` for `Text` and `Interpolator` would probably fix all of that that is fixable. Since Ractive will put multiple text nodes in for interpolators and text with sections in its midst, there's not really a way to re-use 100% of the DOM without some fairly serious mangling of how Ractive renders. I tried to come up with a way to check adjacent text fragments (with whatever may be between them) to use a flag to wait on an update to do the splitting, but I failed.

**Comment by kurdin** [comment]
@evs-chris You right, here is updated example with removed whitespace, http://jsfiddle.net/kxpdgo89/11/ But again, for some reason Reactive adds new line for all inserted values in rendered template and it does not match existed DOM structure. I think this can be fixed. 

I was working on my own implementation for active components with flux jquery and dust templates and I had problems with whitespace as well. So the only way to 100% match rendered template with existed HTML is to render same component with same template on server and bootstrap it on client with the same component code. Ultimately this is a big goal, find easy way to reuse components on server, run it on client and re-render only parts that updated with new data and minimal DOM changes. Again, mostly this is important for public websites with interactive pages, where you want to deliver initial HTML state as soon as you can and maybe not be necessary for small widgets or big dedicated Single Page Applications.

**Comment by evs-chris** [comment]
Random thought/question: could we perhaps have a mode for `toHTML` that put an empty CDATA section (or some other non-visible mcguffin) between HTML text nodes such that the enhancement process could use as extricable text node delimiters? Would doing so be beneficial or detrimental, since presumably, the demarcation would need to be removed during enhancement?

**Comment by monoblaine** [comment]
Is there any way (event, logging, etc.) to test whether progressive enhancement is functioning normally? I'd like to find out if/why Ractive had to touch the dom when the page is loaded.

**Comment by fskreuz** [comment]
@monoblaine You can probably do this with just the dev tools https://developer.chrome.com/devtools/docs/dom-and-styles#setting-dom-breakpoints. By default, Ractive instances replace anything under `el`. That would mean *removal* of the existing nodes, which you can check by doing `Break on > Node removal`.

**Comment by monoblaine** [comment]
@fskreuz Thank you!

Right-clicking on the node and selecting `Break on... > Subtree modifications` is what I needed.

**Comment by monoblaine** [comment]
I think the progressive enhancement feature has some problems regarding triple mustaches.

The following example (http://jsfiddle.net/monoblaine/h516ecvz/1/) runs fine:

```js
setTimeout(() => new Ractive({
    el: 'ractivable-el',
    template: '{{txt}}',
    enhance: true,
    data: { txt: 'Hello!' }
}), 6000);
```

This one (http://jsfiddle.net/monoblaine/tqhs0bmb/1/), however, unnecessarily recreates the text node: (not the best example for triple mustashes, I know)

```js
setTimeout(() => new Ractive({
    el: 'ractivable-el',
    template: '{{{txt}}}',
    enhance: true,
    data: { txt: 'Hello!' }
}), 6000);
```

**Comment by fskreuz** [comment]
@monoblaine I think you should file that as another issue so we can track that separately should that be a bug.

**Comment by monoblaine** [comment]
(Sorry to bump this again) As for the whitespace issue, I'm using [a fork of `gulp-ractive`](https://github.com/monoblaine/gulp-ractive). I've created a gulp task to watch my template files: When a template file is modified, the gulp task reads that file, removes whitespaces that start with a new line (`\r?\n\s*`), compiles the template and saves it. It works beautifully (the adjacent text nodes issue and the triple mustaches are some exceptions): The html created on the server (I'm using ClearScript) perfectly matches the one created on the client and Ractive does not touch anything on the client.
